library/psa_crypto_aead.c

```diff
@@ -0,0 +1,358 @@
+/*
+ *  PSA AEAD entry points
+ */
+/*
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "common.h"
+
+#if defined(MBEDTLS_PSA_CRYPTO_C)
+
+#include "psa_crypto_aead.h"
+#include "psa_crypto_core.h"
+
+#include "mbedtls/ccm.h"
+#include "mbedtls/chachapoly.h"
+#include "mbedtls/cipher.h"
+#include "mbedtls/gcm.h"
+
+typedef struct
+{
+    const mbedtls_cipher_info_t *cipher_info;
+    union
+    {
+        unsigned dummy; /* Make the union non-empty even with no supported algorithms. */
+#if defined(MBEDTLS_CCM_C)
+        mbedtls_ccm_context ccm;
+#endif /* MBEDTLS_CCM_C */
+#if defined(MBEDTLS_GCM_C)
+        mbedtls_gcm_context gcm;
+#endif /* MBEDTLS_GCM_C */
+#if defined(MBEDTLS_CHACHAPOLY_C)
+        mbedtls_chachapoly_context chachapoly;
+#endif /* MBEDTLS_CHACHAPOLY_C */
+    } ctx;
+    psa_algorithm_t core_alg;
```

Comment:
Might as well clean this up while at it. I don't believe it is necessary to keep track of `core_alg` in the context, since all call trees contain the algorithm.

Reply:
But we will probably need it for multi-part operation.

Reply:
I'm not going to block the PR over this, but for the record I'll say that I think that if multi-part isn't going to be implemented soon than it would still make sense to remove the unused elements until the implementation is being done. Having structs with unreferenced/unused items is a bit icky.

Reply:
[Multipart AEAD is just around the corner](https://github.com/ARMmbed/mbedtls/pull/4338).

Reply:
Thanks, I hadn't seen that yet. I guess it deserves an explicit link to this PR?

Reply:
Given that this PR looks almost ready but multipart AEAD is still in a draft state, I expect that @paul-elliott-arm will rebase multipart AEAD once driver delegation is merged.